stackslib/src/net/p2p.rs

```diff
@@ -5430,58 +5427,100 @@ impl PeerNetwork {
-            .last_key_value()
-            .map(|(rc, _)| rc.saturating_add(1))
-            .unwrap_or(0);
-        let mut new_agg_pubkeys: Vec<_> = (next_agg_pubkey_rc..=sort_tip_rc)
-            .filter_map(|key_rc| {
-                let ih = sortdb.index_handle(&tip_sn.sortition_id);
-                let agg_pubkey_opt = if self.get_current_epoch().epoch_id < StacksEpochId::Epoch25 {
-                    None
-                } else {
-                    test_debug!(
-                        "Try to get aggregate public key for reward cycle {}",
-                        key_rc
+
+        let prev_rc = cur_rc.saturating_sub(1);
+
+        // keyed by both rc and sortition ID in case there's a bitcoin fork -- we'd want the
+        // canonical reward set to be loaded
+        let cur_rc_sortition_id = sortdb
+            .get_prepare_phase_start_sortition_id_for_reward_cycle(&tip_sn.sortition_id, cur_rc)?;
+        let prev_rc_sortition_id = sortdb
+            .get_prepare_phase_start_sortition_id_for_reward_cycle(&tip_sn.sortition_id, prev_rc)?;
+
+        for (rc, sortition_id) in [
+            (prev_rc, prev_rc_sortition_id),
+            (cur_rc, cur_rc_sortition_id),
+        ]
+        .into_iter()
+        {
+            if let Some(sort_id) = self.current_reward_set_ids.get(&rc) {
+                if sort_id == &sortition_id {
+                    continue;
+                }
+            }
+            let Ok((reward_cycle_info, reward_cycle_sort_id)) = sortdb
+                .get_preprocessed_reward_set_for_reward_cycle(&tip_sn.sortition_id, rc)
+                .map_err(|e| {
+                    warn!(
+                        "Failed to load reward set for cycle {} ({}): {:?}",
+                        rc, &sortition_id, &e
                     );
-                    NakamotoChainState::load_aggregate_public_key(
-                        sortdb,
-                        &ih,
-                        chainstate,
-                        self.burnchain.reward_cycle_to_block_height(key_rc),
-                        &stacks_tip_block_id,
-                        false,
+                    e
+                })
+            else {
+                // NOTE: this should never be reached
+                continue;
+            };
+            if !reward_cycle_info.is_reward_info_known() {
+                // haven't yet processed the anchor block, so don't store
+                test_debug!("Reward cycle info for cycle {} at sortition {} expects the PoX anchor block, so will not cache", rc, &reward_cycle_sort_id);
+                continue;
+            }
+
+            test_debug!(
+                "Reward cycle info for cycle {} at sortition {} is {:?}",
+                rc,
+                &reward_cycle_sort_id,
+                &reward_cycle_info
+            );
+            self.current_reward_sets.insert(rc, reward_cycle_info);
+            self.current_reward_set_ids.insert(rc, reward_cycle_sort_id);
```

Comment:
Would `current_reward_sets: BTreeMap<u64, (SortitionId, RewardCycleInfo)>` work for tracking the information? Or is there a case where `current_reward_sets.get(x).is_some() && current_reward_set_ids.get(x).is_none()`?